### ISSUE TITLE
Files deleted after the streaming finishes

### ISSUE THREAD

**Issue by tradingstratagem** [illuspas/Node-Media-Server#341]
I noticed that all the files are deleted after the streaming has been stopped.
The code is part of the core and it doesn't seem to be avoided.

Can anyone give me some suggestions on how to avoid this? (Ok I can fork and change this behaviours, but I'ld prefer to avoid this)

**Comment by phprad** [comment]
You can remove the option to delete the segments from the task
```
{  
		app: 'live',  
		ac: 'aac',  
		hls: true,  
		hlsFlags: '[hls_time=10:hls_list_size=0_**:hls_flags=delete_segments**_]',
		dash: false,  
		dashFlags: '[f=dash:window_size=3:extra_window_size=5]'  
	  }
```

**Comment by tradingstratagem** [comment]
it doesn't work: here is the code that runs when the process exits:

this.ffmpeg_exec.on('close', (code) => {
      Logger.log('[Transmuxing end] ' + this.conf.streamPath);
      this.emit('end');
      fs.readdir(ouPath, function (err, files) {
        if (!err) {
          files.forEach((filename) => {
            if (filename.endsWith('.ts')
              || filename.endsWith('.m3u8')
              || filename.endsWith('.mpd')
              || filename.endsWith('.m4s')
              || filename.endsWith('.tmp')) {
              fs.unlinkSync(ouPath + '/' + filename);
            }
          })
        }
      });
    });

**Comment by RupamShaw** [comment]
yes, all files deleted. I can see during streaming files (index.m3u8 and index0.ts , index1.ts...) creates and after transmusing all files deleted.

**Comment by Viruss98** [comment]
mp4: true,
mp4Flags: '[movflags=faststart]'
add two lines of code to your config in tasks: 
eg: 
trans: {
            ffmpeg: '/usr/bin/ffmpeg',
            tasks: [
                {
                    ...yours,
                    mp4: true,
                    mp4Flags: '[movflags=faststart]'
                }
            ]
        }

**Comment by Viruss98** [comment]
> I noticed that all the files are deleted after the streaming has been stopped.
> The code is part of the core and it doesn't seem to be avoided.
> 
> Can anyone give me some suggestions on how to avoid this? (Ok I can fork and change this behaviours, but I'ld prefer to avoid this)

@tradingstratagem 
mp4: true,
mp4Flags: '[movflags=faststart]'
add two lines of code to your config in tasks:
eg:
trans: {
ffmpeg: '/usr/bin/ffmpeg',
tasks: [
{
...yours,
mp4: true,
mp4Flags: '[movflags=faststart]'
}
]
}

**Comment by perinazzoo** [comment]
> I noticed that all the files are deleted after the streaming has been stopped.
> The code is part of the core and it doesn't seem to be avoided.
> 
> Can anyone give me some suggestions on how to avoid this? (Ok I can fork and change this behaviours, but I'ld prefer to avoid this)

I know that its been a long time since the last reply but I want to know: did you find any solutions for this? I'm stuck with the same problem because the player I'm using in the front end will wait for "EXT-X-ENDLIST" in the index.m3u8 but it's not included at any time.

**Comment by joepsyko** [comment]
No, at the end I gave up and I wrote my own implementation

**Comment by Viruss98** [comment]
I can save file as  .mp4 video after stream finish, my stream is still using .m3u8 while stream run. This is my config for my company's project, it run without error 
![Screenshot_20201007-192545~2](https://user-images.githubusercontent.com/29653304/95331001-8ce6f500-08d3-11eb-8235-9469ef70c212.png)

**Comment by perinazzoo** [comment]
I found a solution digging the lib and Apple's documentation and I saw the problems:

If you set a hlsFlag `hls_playlist_type=event` it should add the "EXT-X-ENDLIST" line in the end of your index.m3u8 file, but it doesn't happen because:

1. In the end of node_trans_session.js on close listener, the files are deleted instantly like @tradingstratagem said.

2. In the end of node_trans_server.js file, the method **onDonePublish** closes the Transmuxing session and the ffmpeg can't append the "EXT-X-ENDLIST" line.

I was thinking about removing the code that deletes all files at the end of stream and let it delete only the temp files and add this code at the start of streaming, so we can have a clean new streaming folder (this allows the player on the front end to consume the latest live as a video).

About the session closing:
I would like you to give me a light because I have no idea how this works, but I know that if the session is not closed, the line is added.

Best Regards

**Comment by gigabyteservice** [comment]
fs.unlinkSync(ouPath + '/' + filename); I have commented this code, hls files remain as it is bit i can't able to play those files after streaming ends. 

We want hls to be played even after streaming is finished
